<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="robot">

  <!-- CONSTANTS -->
  <xacro:property name="base_length" value="0.4" />
  <xacro:property name="base_width" value="0.2" />
  <xacro:property name="base_height" value="0.1" />
  <xacro:property name="wheel_length" value="0.05" />
  <xacro:property name="wheel_radius" value="0.1" />
  <xacro:property name="camera_link" value="0.05" />
  <xacro:property name="camera_name" value="xtion" />
  <xacro:property name="M_PI" value="3.1415926535897931" />

  
  <!-- BASE LINK -->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="${base_length} ${base_width} ${base_height}"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <box size="${base_length} ${base_width} ${base_height}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="-${base_length/2 - base_height/2} 0 -${base_height/2}" rpy="0 0 0" />
      <geometry>
        <sphere radius="${base_height/2}" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="-${base_length/2 - base_height/2} 0 -${base_height/2}" rpy="0 0 0" />
      <geometry>
        <sphere radius="${base_height/2}" />
      </geometry>
    </collision>
    <inertial>
      <mass value="0.6"/>
      <inertia ixx="1e-3" ixy="0.0" ixz="0.0" iyy="1e-3" iyz="0.0" izz="1e-3"/>
    </inertial>
  </link>
  <gazebo reference="base_link">
    <mu1>0.0</mu1>
    <mu2>0.0</mu2>
    <material>Gazebo/Black</material>
  </gazebo>

  <!-- LEFT WHEEL -->
  <link name="left_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.5707 0 0" />
      <geometry>
        <cylinder length="${base_height/2}" radius="${base_height}" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="1.5707 0 0" />
      <geometry>
        <cylinder length="${base_height/2}" radius="${base_height}" />
      </geometry>
    </collision>
    <inertial>
      <mass value="0.6"/>
      <inertia ixx="1e-3" ixy="0.0" ixz="0.0" iyy="1e-3" iyz="0.0" izz="1e-3"/>
    </inertial>
  </link>
  <gazebo reference="left_wheel">
    <mu1 value="200.0"/>
    <mu2 value="100.0"/>
    <kp value="10000000.0" />
    <kd value="1.0" />
    <material>Gazebo/Grey</material>
  </gazebo>

  <!-- LEFT WHEEL JOINT -->
  <joint name="left_wheel_joint" type="continuous">
    <parent link="base_link" />
    <child  link="left_wheel" />
    <origin xyz="0.1 0.13 0" />
    <axis rpy="0 0 0" xyz="0 1 0"/>
  </joint>

  <!-- LEFT WHEEL JOINT TRANSMISSION -->
  <transmission name="left_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$left_wheel_joint_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="left_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <!-- RIGHT WHEEL -->
  <link name="right_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="1.5707 0 0" />
      <geometry>
        <cylinder length="${base_height/2}" radius="${base_height}" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="1.5707 0 0" />
      <geometry>
        <cylinder length="${base_height/2}" radius="${base_height}" />
      </geometry>
    </collision>
    <inertial>
      <mass value="0.6"/>
      <inertia ixx="1e-3" ixy="0.0" ixz="0.0" iyy="1e-3" iyz="0.0" izz="1e-3"/>
    </inertial>
  </link>
  <gazebo reference="right_wheel">
    <mu1 value="200.0"/>
    <mu2 value="100.0"/>
    <kp value="10000000.0" />
    <kd value="1.0" />
    <material>Gazebo/Grey</material>
  </gazebo>

  <!-- RIGHT WHEEL JOINT -->
  <joint name="right_wheel_joint" type="continuous">
    <parent link="base_link" />
    <child  link="right_wheel" />
    <origin xyz="0.1 -0.13 0" />
    <axis rpy="0 0 0" xyz="0 1 0"/>
  </joint>

  <!-- RIGHT WHEEL JOINT TRANSMISSION -->
  <transmission name="right_wheel_joint_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$right_wheel_joint_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="right_wheel_joint">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <!-- LASER -->
  <link name="laser_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.1 0.1 0.1" />
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://semantic_maps/meshes/hokuyo.dae" />
      </geometry>
    </visual>
    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <!-- LASER JOINT -->
  <joint name="laser_joint" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="0.1 0 0.1" rpy="0 0 0" />
    <parent link="base_link" />
    <child  link="laser_link" />
  </joint>

  <!-- HOKUYO SENSOR -->
  <gazebo reference="laser_link">
    <sensor type="ray" name="head_hokuyo_sensor">
      <pose>0 0 0 0 0 0</pose>
      <visualize>true</visualize>
      <update_rate>40</update_rate>
      <ray>
        <scan>
          <horizontal>
            <samples>1024</samples>
            <resolution>1</resolution>
            <min_angle>-1.57</min_angle>
            <max_angle>1.57</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.02</min>
          <max>10.0</max>
          <resolution>0.1</resolution>
        </range>
        <noise>
          <type>gaussian</type>
          <mean>0.0</mean>
          <stddev>0.0</stddev>
        </noise>
      </ray>
      <plugin name="gazebo_ros_head_hokuyo_controller" filename="libgazebo_ros_laser.so">
        <topicName>/scan</topicName>
        <frameName>laser_link</frameName>
      </plugin>
    </sensor>
  </gazebo>

  <!-- CAMERA -->
  <link name="camera_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="${camera_link} ${camera_link} ${camera_link}"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="${camera_link} ${camera_link} ${camera_link}"/>
      </geometry>
      <material name="red"/>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>
  

  <!-- CAMERA JOINT -->
  <joint name="camera_joint" type="continuous">
    <axis xyz="0 1 0" />
    <origin xyz="0 0 1.0" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="camera_link"/>
  </joint>
<!--
  <gazebo reference="camera_joint">
    <preserveFixedJoint>true</preserveFixedJoint>
  </gazebo>"
-->  

  <!-- XTION SENSOR -->
  <gazebo reference="camera_link">
    <sensor type="depth" name="rgbd_camera">
      <always_on>true</always_on>
      <update_rate>20.0</update_rate>
      <camera>
        <horizontal_fov>${60.0*M_PI/180.0}</horizontal_fov>
        <image>
          <width>320</width>
          <height>240</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.02</near>
          <far>8.0</far>
        </clip>
      </camera>
      <plugin name="rgdb_camera_controller" filename="libgazebo_ros_openni_kinect.so">
        <cameraName>camera</cameraName>
          <alwaysOn>true</alwaysOn>
          <updateRate>10</updateRate>
          <imageTopicName>rgb/image_raw</imageTopicName>
          <depthImageTopicName>depth/image_raw</depthImageTopicName>
          <pointCloudTopicName>depth/points</pointCloudTopicName>
          <cameraInfoTopicName>rgb/camera_info</cameraInfoTopicName>
          <depthImageCameraInfoTopicName>depth/camera_info</depthImageCameraInfoTopicName>
          <frameName>camera_depth_optical_frame</frameName>
          <baseline>0.1</baseline>
          <distortion_k1>0.0</distortion_k1>
          <distortion_k2>0.0</distortion_k2>
          <distortion_k3>0.0</distortion_k3>
          <distortion_t1>0.0</distortion_t1>
          <distortion_t2>0.0</distortion_t2>
          <pointCloudCutoff>0.4</pointCloudCutoff>
      </plugin>
    </sensor>
  </gazebo>
  

  <!-- Gazebo plugin for ROS Control -->
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/</robotNamespace>
    </plugin>
  </gazebo>

</robot>
